File: client/src/services/productService.js
const API_URL = 'http://localhost:5000/api';

const mapProduct = (p) => {
  if (!p) return null;
  return {
    ...p,
    id: p._id || p.id,
  };
};

/**
 * Fetch products from the backend with optional filters.
 * @param {Object} filters - Filter criteria (category, search, minPrice, maxPrice, sizes, colors, sort).
 * @returns {Promise<Array>} List of products.
 */
export const fetchProducts = async (filters = {}) => {
  try {
    const queryParams = new URLSearchParams();
    Object.entries(filters).forEach(([key, val]) => {
      if (val !== undefined && val !== null && val !== '') {
        queryParams.append(key, val);
      }
    });

    const res = await fetch(`${API_URL}/products?${queryParams.toString()}`);
    if (!res.ok) throw new Error('Failed to fetch products');
    const data = await res.json();
    return data.map(mapProduct);
  } catch (error) {
    console.error('Error fetching products:', error);
    return [];
  }
};

/**
 * Fetch a single product by its ID.
 * @param {string} id - Product ID.
 * @returns {Promise<Object|null>} Product details.
 */
export const fetchProductById = async (id) => {
  try {
    const res = await fetch(`${API_URL}/products/${id}`);
    if (!res.ok) throw new Error('Product not found');
    const data = await res.json();
    return mapProduct(data);
  } catch (error) {
    console.error(`Error fetching product by ID ${id}:`, error);
    return null;
  }
};

/**
 * Fetch related products of a given product.
 * @param {string} category - Product category.
 * @param {string} excludeId - Current product ID to exclude.
 * @returns {Promise<Array>} List of related products.
 */
export const fetchRelatedProducts = async (category, excludeId) => {
  try {
    const products = await fetchProducts({ category });
    return products.filter((p) => p.id !== excludeId).slice(0, 4);
  } catch (error) {
    console.error('Error fetching related products:', error);
    return [];
  }
};

File: mobile/lib/services/product_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/product_model.dart';
import '../core/constants/api_constants.dart';

/// ProductService — mirrors client/src/services/productService.js
class ProductService {
  static final ProductService _instance = ProductService._internal();
  factory ProductService() => _instance;
  ProductService._internal();

  /// Fetch products with optional filters
  /// Mirrors: fetchProducts(filters)
  Future<List<ProductModel>> fetchProducts({
    String? category,
    String? search,
    double? minPrice,
    double? maxPrice,
    List<String>? sizes,
    List<String>? colors,
    String? sort,
  }) async {
    try {
      final queryParams = <String, String>{};
      if (category != null && category.isNotEmpty) queryParams['category'] = category;
      if (search != null && search.isNotEmpty) queryParams['search'] = search;
      if (minPrice != null) queryParams['minPrice'] = minPrice.toString();
      if (maxPrice != null) queryParams['maxPrice'] = maxPrice.toString();
      if (sizes != null && sizes.isNotEmpty) queryParams['sizes'] = sizes.join(',');
      if (colors != null && colors.isNotEmpty) queryParams['colors'] = colors.join(',');
      if (sort != null && sort.isNotEmpty) queryParams['sort'] = sort;

      final uri = Uri.parse(ApiConstants.products).replace(queryParameters: queryParams);
      final response = await http.get(uri).timeout(const Duration(seconds: 15));

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((json) => ProductModel.fromJson(json)).toList();
      }
      // ignore: avoid_print
      print('[ProductService] fetchProducts failed: ${response.statusCode} ${response.body.substring(0, response.body.length.clamp(0, 200))}');
      return [];
    } catch (e) {
      // ignore: avoid_print
      print('[ProductService] fetchProducts error: $e');
      return [];
    }
  }

  /// Fetch a single product by ID
  /// Mirrors: fetchProductById(id)
  Future<ProductModel?> fetchProductById(String id) async {
    try {
      final response = await http
          .get(Uri.parse(ApiConstants.productById(id)))
          .timeout(const Duration(seconds: 15));

      if (response.statusCode == 200) {
        return ProductModel.fromJson(jsonDecode(response.body));
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  /// Fetch related products in same category, excluding current product
  /// Mirrors: fetchRelatedProducts(category, excludeId)
  Future<List<ProductModel>> fetchRelatedProducts(String category, String excludeId) async {
    final products = await fetchProducts(category: category);
    return products.where((p) => p.id != excludeId).take(4).toList();
  }

  /// Fetch hero config
  Future<Map<String, dynamic>?> fetchHeroConfig() async {
    try {
      final response = await http
          .get(Uri.parse(ApiConstants.heroConfig))
          .timeout(const Duration(seconds: 10));
      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      }
      return null;
    } catch (e) {
      return null;
    }
  }
}
